app/validators/rules/rule_validator.py

```diff
@@ -0,0 +1,215 @@
+            function_to_map_over_arguments
+        )
+
+        if SELF_REFERENCE_KEY not in arguments_for_non_map_operators:
+            self.add_error(
+                self.MAP_OPERATOR_WITHOUT_SELF_REFERENCE, rule=operator["map"][0]
+            )
+
+    def _get_flattened_arguments_for_non_map_operators(self, arguments):
+        """
+        Recursively fetch all the arguments for all non `map` operators.
+
+        The `map` operator is checked explicitly.
+        """
+        non_operator_arguments = []
+        for argument in arguments:
+            if isinstance(argument, dict) and any(
+                operator in argument
+                for operator in ALL_OPERATORS
+                if operator != Operator.MAP
+            ):
+                for operands in argument.values():
+                    non_operator_arguments += self._get_flattened_arguments_for_non_map_operators(
+                        operands
+                    )
+            else:
+                non_operator_arguments.append(argument)
+
+        return non_operator_arguments
+
+    def _validate_date_operator(self, operator):
+        """
+        Validates that when an answer value source is used it is a date
+        """
+        first_argument = operator["date"][0]
+        if (
+            isinstance(first_argument, dict)
+            and first_argument.get("source") == "answers"
+            and self.questionnaire_schema.get_answer(first_argument["identifier"])[
+                "type"
+            ]
+            not in ["Date", "MonthYearDate", "YearDate"]
+        ):
+            self.add_error(
+                self.DATE_OPERATOR_REFERENCES_NON_DATE_ANSWER,
+                value_source=first_argument,
+            )
+
+    def _validate_count_operator(self, operator):
+        """
+        Validates that an answer value source within a count operator is of type Checkbox
+        """
+        first_argument = operator["count"][0]
+        if (
+            isinstance(first_argument, dict)
+            and first_argument.get("source") == "answers"
+            and self.questionnaire_schema.get_answer(first_argument["identifier"])[
+                "type"
+            ]
+            != "Checkbox"
+        ):
+            self.add_error(
+                self.COUNT_OPERATOR_REFERENCES_NON_CHECKBOX_ANSWER,
+                value_source=first_argument,
+            )
+
+    def _validate_options(self, rules, operator_name):
+        """
+        Validates that answer options referenced in a rule exist
+        """
+        values = []
+        option_values = []
+        for argument in rules[operator_name]:
+            if isinstance(argument, dict) and argument.get("source") == "answers":
+                option_values = self.questionnaire_schema.answer_id_to_option_values_map.get(
+                    argument["identifier"]
+                )
+            else:
+                values = argument if isinstance(argument, list) else [argument]
```

Comment:
`option_values` and `values` will only have last values of `argument` ? as they are getting overwritten

Reply:
This was copied from the existing file, but I see what you mean. However, this works because 2 operators can never be at the same level, it always has to be nested if you have more than 1 operator. So technically  this is always a dict of length one, which can have 2 arguments, i.e:
```py
{'==': [{'source': 'answers', 'identifier': 'string-answer'}, 'Maybe']}
```

This struct would be invalid:
```py
{
	'==': [{'source': 'answers', 'identifier': 'string-answer'}, 'Maybe'],
	'!=': [{'source': 'answers', 'identifier': 'string-answer'}, 'Maybe'],
}
```

`values` should never be populated when the `else` clause is reached if the schema is valid.